Preparation for transaction sequence testing.

jonathanknowles opened the pull request:
## Issue Number

ADP-1783

## Summary

This PR adds the following generator functions to `Test.QuickCheck.Extra`:

- **`partitionList`**
    partitions a list into a list of sublists of random lengths.
    
- `selectMapEntry`
    selects an entry at random from a `Map`.

- `selectMapEntries`
    selects up to a given number of entries at random from a `Map`.

## Motivation

- **`selectMapEntries`**
    can be used to randomly select one or more UTxO entries to act as inputs when generating transactions.

- **`partitionList`**
    can be used to transform lists of transactions into arbitrary lists of blocks, where each block can contain zero, one, or many transactions.

jonathanknowles committed:
Add generator function `partitionList`.

jonathanknowles committed:
Add generator function `mapEntry`.

jonathanknowles committed:
Add generator function `mapEntries`.

sevanspowell commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -249,6 +252,27 @@ partitionList (x, y) =
+--
+-- Returns 'Nothing' if (and only if) the given map is empty.
+--
+mapEntry :: forall k v. Ord k => Map k v -> Gen (Maybe ((k, v), Map k v))
```
I think this family of functions should be called `getMapEntry` or similar. `mapEntry` is a bit similar to `map` (i.e `fmap`), which has the connotation of "mapping over" something. I think if it was named `fmapEntry` this erroneous connotation would be more obvious.

Just an opinion, not a blocker.

jonathanknowles replied:
Thanks @sevanspowell, this is great feedback.

I've changed the names in 44df8f839404e19c3223600d525f00f5a1d21015.

jonathanknowles committed:
Rename `map{Entry,Entries}` to `selectMap{Entry,Entries}`.

In response to review feedback and subsequent 1:1 discussion with Sam:

https://github.com/input-output-hk/cardano-wallet/pull/3295#pullrequestreview-986991172
